samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -132,71 +136,72 @@ private void validateConfig() {
    */
   /* package private */ void calculatePartitions(JobGraph jobGraph) {
     // calculate the partitions for the input streams of join operators
-    calculateJoinInputPartitions(jobGraph, config);
+    calculateJoinInputPartitions(jobGraph, streamConfig);
 
     // calculate the partitions for the rest of intermediate streams
-    calculateIntStreamPartitions(jobGraph, config);
+    calculateIntermediateStreamPartitions(jobGraph, config);
 
     // validate all the partitions are assigned
-    validatePartitions(jobGraph);
+    validateIntermediateStreamPartitions(jobGraph);
   }
 
   /**
-   * Fetch the partitions of source/sink streams and update the StreamEdges.
+   * Fetch the partitions of input/output streams and update the corresponding StreamEdges.
    * @param jobGraph {@link JobGraph}
    * @param streamManager the {@link StreamManager} to interface with the streams.
    */
-  /* package private */ static void updateExistingPartitions(JobGraph jobGraph, StreamManager streamManager) {
+  /* package private */ static void fetchInputAndOutputStreamPartitions(JobGraph jobGraph, StreamManager streamManager) {
     Set<StreamEdge> existingStreams = new HashSet<>();
-    existingStreams.addAll(jobGraph.getSources());
-    existingStreams.addAll(jobGraph.getSinks());
-
-    Multimap<String, StreamEdge> systemToStreamEdges = HashMultimap.create();
-    // group the StreamEdge(s) based on the system name
-    existingStreams.forEach(streamEdge -> {
-        SystemStream systemStream = streamEdge.getSystemStream();
-        systemToStreamEdges.put(systemStream.getSystem(), streamEdge);
-      });
-    for (Map.Entry<String, Collection<StreamEdge>> entry : systemToStreamEdges.asMap().entrySet()) {
-      String systemName = entry.getKey();
-      Collection<StreamEdge> streamEdges = entry.getValue();
-      Map<String, StreamEdge> streamToStreamEdge = new HashMap<>();
-      // create the stream name to StreamEdge mapping for this system
-      streamEdges.forEach(streamEdge -> streamToStreamEdge.put(streamEdge.getSystemStream().getStream(), streamEdge));
-      // retrieve the partition counts for the streams in this system
-      Map<String, Integer> streamToPartitionCount = streamManager.getStreamPartitionCounts(systemName, streamToStreamEdge.keySet());
-      // set the partitions of a stream to its StreamEdge
-      streamToPartitionCount.forEach((stream, partitionCount) -> {
-          streamToStreamEdge.get(stream).setPartitionCount(partitionCount);
-          log.info("Partition count is {} for stream {}", partitionCount, stream);
-        });
+    existingStreams.addAll(jobGraph.getInputStreams());
+    existingStreams.addAll(jobGraph.getOutputStreams());
+
```

Comment:
We are effectively un-nesting the lookup in this map by one-level. Previously, the map of [streamName -> streamEdge] was per-system. It appears that we have changed this to a single-map of [streamName -> streamEdge]..

Do you think it could present issues when we read the same streamName from two different systems? As a concrete example at LinkedIn: imagine reading the same stream - PageViewEvent from both the "tracking" and "aggregate-tracking" Kafka clusters. We would like the "system-level" scope to be considered and the streams treated as 2 distinct ones - each with their own StreamEdges.

Reply:
I wasn't aware of that. Thanks for the catch. I have changed the logic accordingly.

Reply:
If you still prefer the top-level map style, you could change the map from system-stream -> streamEdge. I'll leave that entirely up-to you.

Reply:
I did consider that but this would entail creating a system-stream object for every stream after calling `streamManager.getStreamPartitionCounts()` to index into the top-level map.

I think one of the primary root causes of the complexity present in this method is in fact due to `streamManager.getStreamPartitionCounts()``s signature.